lib/Service/IMipService.php

```diff
@@ -0,0 +1,168 @@
+use OCA\Mail\Db\MessageMapper;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Model\IMAPMessage;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\Calendar\IManager;
+use Psr\Log\LoggerInterface;
+
+class IMipService {
+	private AccountService $accountService;
+	private IManager $calendarManager;
+	private LoggerInterface $logger;
+	private MailboxMapper $mailboxMapper;
+	private MailManager $mailManager;
+	private MessageMapper $messageMapper;
+
+	public function __construct(
+		AccountService $accountService,
+		IManager $manager,
+		LoggerInterface $logger,
+		MailboxMapper $mailboxMapper,
+		MailManager $mailManager,
+		MessageMapper $messageMapper
+	) {
+		$this->accountService = $accountService;
+		$this->calendarManager = $manager;
+		$this->logger = $logger;
+		$this->mailboxMapper = $mailboxMapper;
+		$this->mailManager = $mailManager;
+		$this->messageMapper = $messageMapper;
+	}
+
+	public function process(): void {
+		$messages = $this->messageMapper->findIMipMessagesAscending();
+		if (empty($messages)) {
+			$this->logger->info('No iMIP messages to process.');
+			return;
+		}
+
+		// Collect all mailboxes in memory
+		// possible perf improvement - make this one IN query
+		// and JOIN with accounts table
+		// although this might not make much of a difference
+		// since there are very few messages to process
+		$mailboxIds = array_unique(array_map(function (Message $message) {
+			return $message->getMailboxId();
+		}, $messages));
+
+		$mailboxes = array_map(function (int $mailboxId) {
+			try {
+				return $this->mailboxMapper->findById($mailboxId);
+			} catch (DoesNotExistException | ServiceException $e) {
+				return null;
+			}
+		}, $mailboxIds);
+
+		// Collect all accounts in memory
+		$accountIds = array_unique(array_map(function (Mailbox $mailbox) {
+			return $mailbox->getAccountId();
+		}, $mailboxes));
+
+		$accounts = array_combine($accountIds, array_map(function (int $accountId) {
+			try {
+				return $this->accountService->findById($accountId);
+			} catch (DoesNotExistException $e) {
+				return null;
+			}
+		}, $accountIds));
+
+		/** @var Mailbox $mailbox */
+		foreach ($mailboxes as $mailbox) {
+			/** @var Account $account */
+			$account = $accounts[$mailbox->getAccountId()];
+			$filteredMessages = array_filter($messages, function ($message) use ($mailbox) {
+				return $message->getMailboxId() === $mailbox->getId();
+			});
+
+			if (empty($filteredMessages)) {
+				continue;
+			}
+
+			// Check for accounts or mailboxes that no longer exist,
+			// no processing for drafts, sent items, junk or trash
+			if ($account === null
+				|| $account->getMailAccount()->getTrashMailboxId() === $mailbox->getId()
+				|| $account->getMailAccount()->getSentMailboxId() === $mailbox->getId()
+				|| $account->getMailAccount()->getDraftsMailboxId() === $mailbox->getId()
+				|| $mailbox->isSpecialUse(\Horde_Imap_Client::SPECIALUSE_ARCHIVE)
+			) {
+				array_map(function ($message) {
```

Comment:
That's a bit weird. It works because objects are passed by reference. I would prefer to store the new array or merge the line with `$this->messageMapper->updateImipData`